### PR TITLE
4: SSG Markdown Index Page

### PR DESCRIPTION
**Pull request by jordan-bravo**
This update adds functionality to parse markdown files in the posts directory and render them as a list on the index page.

Issue #4

### CODE REVIEW

**Comment by sbddesign** [comment]
@jordan-bravo I'm getting this error on the Netlify build:
  
  ```
  7:16:05 PM: Failed during stage 'building site': Build script returned non-zero exit code: 2 (https://ntl.fyi/exit-code-2)
7:15:32 PM: build-image version: ad28fb91fae632ef7f2e4d659a892d68ee2681a6 (focal)
7:15:32 PM: buildbot version: 751c8de2ba48c52549712392ca0244c53d39c3c1
7:15:32 PM: Fetching cached dependencies
7:15:32 PM: Starting to download cache of 779.2MB
7:15:34 PM: Finished downloading cache in 1.989s
7:15:34 PM: Starting to extract cache
7:15:38 PM: Finished extracting cache in 4.002s
7:15:38 PM: Finished fetching cache in 6.079s
7:15:38 PM: Starting to prepare the repo for build
7:15:38 PM: Preparing Git Reference pull/9/head
7:15:40 PM: Starting to install dependencies
7:15:40 PM: Python version set to 3.8
7:15:40 PM: Attempting Ruby version 2.7.2, read from environment
7:15:41 PM: Using Ruby version 2.7.2
7:15:41 PM: Started restoring cached go cache
7:15:41 PM: Finished restoring cached go cache
7:15:41 PM: go version go1.19.11 linux/amd64
7:15:42 PM: Using PHP version 8.0
7:15:42 PM: Started restoring cached Node.js version
7:15:43 PM: Finished restoring cached Node.js version
7:15:43 PM: v18.17.0 is already installed.
7:15:44 PM: Now using node v18.17.0 (npm v9.6.7)
7:15:44 PM: Enabling Node.js Corepack
7:15:44 PM: Started restoring cached build plugins
7:15:44 PM: Finished restoring cached build plugins
7:15:44 PM: Started restoring cached corepack dependencies
7:15:44 PM: Finished restoring cached corepack dependencies
7:15:44 PM: Started restoring cached yarn cache
7:15:45 PM: Finished restoring cached yarn cache
7:15:45 PM: No yarn workspaces detected
7:15:45 PM: Started restoring cached node modules
7:15:45 PM: Finished restoring cached node modules
7:15:46 PM: Installing npm packages using Yarn version 1.22.19
7:15:46 PM: yarn install v1.22.19
7:15:46 PM: [1/4] Resolving packages...
7:15:46 PM: [2/4] Fetching packages...
7:15:47 PM: [3/4] Linking dependencies...
7:15:48 PM: [4/4] Building fresh packages...
7:15:48 PM: Done in 2.08s.
7:15:48 PM: npm packages installed using Yarn
7:15:49 PM: Install dependencies script success
7:15:49 PM: Starting build script
7:15:49 PM: Detected 1 framework(s)
7:15:49 PM: next at version 13.4.9
7:15:49 PM: Section completed: initializing
7:15:51 PM: ​
7:15:51 PM: Netlify Build                                                 
7:15:51 PM: ────────────────────────────────────────────────────────────────
7:15:51 PM: ​
7:15:51 PM: ❯ Version
7:15:51 PM:   @netlify/build 29.17.1
7:15:51 PM: ​
7:15:51 PM: ❯ Flags
7:15:51 PM:   baseRelDir: true
7:15:51 PM:   buildId: 64bc6312ce3e9a0008b02387
7:15:51 PM:   deployId: 64bc6312ce3e9a0008b02389
7:15:51 PM: ​
7:15:51 PM: ❯ Current directory
7:15:51 PM:   /opt/build/repo
7:15:51 PM: ​
7:15:51 PM: ❯ Config file
7:15:51 PM:   No config file was defined: using default values.
7:15:51 PM: ​
7:15:51 PM: ❯ Context
7:15:51 PM:   deploy-preview
7:15:51 PM: ​
7:15:51 PM: ❯ Using Next.js Runtime - v4.39.1
7:15:52 PM: ​
7:15:52 PM: @netlify/plugin-nextjs (onPreBuild event)                     
7:15:52 PM: ────────────────────────────────────────────────────────────────
7:15:52 PM: ​
7:15:52 PM: Next.js cache restored.
7:15:52 PM: Netlify configuration property build.environment.NEXT_PRIVATE_TARGET value changed.
7:15:52 PM: ​
7:15:52 PM: (@netlify/plugin-nextjs onPreBuild completed in 43ms)
7:15:52 PM: ​
7:15:52 PM: Build command from Netlify app                                
7:15:52 PM: ────────────────────────────────────────────────────────────────
7:15:52 PM: ​
7:15:52 PM: $ yarn build
7:15:52 PM: yarn run v1.22.19
7:15:52 PM: $ next build
7:15:53 PM: - info Creating an optimized production build...
7:16:00 PM: - info Compiled successfully
7:16:00 PM: - info Linting and checking validity of types...
7:16:02 PM: Failed to compile.
7:16:02 PM: 
7:16:02 PM: ./app/page.tsx:12:34
7:16:02 PM: Type error: Property 'date' does not exist on type '{ id: string; }'.
7:16:02 PM:   10 |       <Link href=/events>Events</Link>
7:16:02 PM:   11 |       <ul>
7:16:02 PM: > 12 |         {allPostsData.map(({ id, date, title }) => (
7:16:02 PM:      |                                  ^
7:16:02 PM:   13 |           <li className=post key={id}>
7:16:02 PM:   14 |             {title}
7:16:02 PM:   15 |             <br />
7:16:02 PM: error Command failed with exit code 1. (https://ntl.fyi/exit-code-1)
7:16:02 PM: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
7:16:02 PM: ​
7:16:02 PM: build.command failed                                        
7:16:02 PM: ────────────────────────────────────────────────────────────────
7:16:02 PM: ​
7:16:02 PM:   Error message
7:16:02 PM:   Command failed with exit code 1: yarn build (https://ntl.fyi/exit-code-1)
7:16:02 PM: ​
7:16:02 PM:   Error location
7:16:02 PM:   In Build command from Netlify app:
7:16:02 PM:   yarn build
7:16:02 PM: ​
7:16:02 PM:   Resolved config
7:16:02 PM:   build:
7:16:02 PM:     command: yarn build
7:16:02 PM:     commandOrigin: ui
7:16:02 PM:     environment:
7:16:02 PM:       - REVIEW_ID
7:16:02 PM:       - NEXT_PRIVATE_TARGET
7:16:02 PM:     publish: /opt/build/repo/.next
7:16:02 PM:     publishOrigin: ui
7:16:02 PM:   plugins:
7:16:02 PM:     - inputs: {}
7:16:02 PM:       origin: ui
7:16:02 PM:       package: '@netlify/plugin-nextjs'
7:16:02 PM: Caching artifacts
7:16:02 PM: Started saving node modules
7:16:02 PM: Finished saving node modules
7:16:02 PM: Started saving build plugins
7:16:02 PM: Finished saving build plugins
7:16:02 PM: Started saving corepack cache
7:16:02 PM: Finished saving corepack cache
7:16:02 PM: Started saving yarn cache
7:16:04 PM: Finished saving yarn cache
7:16:04 PM: Started saving pip cache
7:16:04 PM: Finished saving pip cache
7:16:04 PM: Started saving emacs cask dependencies
7:16:04 PM: Finished saving emacs cask dependencies
7:16:04 PM: Started saving maven dependencies
7:16:04 PM: Finished saving maven dependencies
7:16:04 PM: Started saving boot dependencies
7:16:04 PM: Finished saving boot dependencies
7:16:04 PM: Started saving rust rustup cache
7:16:04 PM: Finished saving rust rustup cache
7:16:04 PM: Started saving go dependencies
7:16:04 PM: Finished saving go dependencies
7:16:04 PM: Build failed due to a user error: Build script returned non-zero exit code: 2
7:16:04 PM: Failing build: Failed to build site
7:16:05 PM: Finished processing build request in 32.843s
```

**Comment by jordan-bravo** [comment]
The issue is in `/app/page.tsx`:
```ts
12        {allPostsData.map(({ id, date, title }) => (
```

`Property 'date' does not exist on type '{ id: string; }'.`

I'm having trouble figuring out why TypeScript is throwing an error because `allPostsData` should have those properties.  If you log it out, you can see that it's an array of objects:

```ts
[
  {
    id: 'really-cool',
    title: 'Really Cool Topic Title',
    date: '2023-07-21'
  },
  {
    id: 'totally-awesome',
    title: 'Totally Awesome Topic Title',
    date: '2023-06-19'
  },
  {
    id: 'yet-another',
    title: 'Yet Another Topic Title',
    date: '2023-05-28'
  }
]
```

I'll take another look later but feel free to work on it if you can figure it out.

**Review by sbddesign — APPROVED** [review]
@jordan-bravo I fixed and approved it. Here's the issue: while all the markdown files do have a title and date, there was nothing built in to handle hypotehtical situations where a markdown file is missing one or both of those properties. `yarn build` is a little stricter than `yarn dev` about these things. To demonstrate, I made a file called "missing info" and sure enough, the build fails for these. I added in a default title and date. Maybe  there is a more elegant solution for the future, but that works for now.

File: utils/posts.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    // Use gray-matter to parse the post metadata section
+    const matterResult = matter(fileContents)
+
+    // Combine the data with the id
+    return {
+      id,
+      title: "Meetup Event",
+      date: "2009-01-03 00:00:00",
+      ...matterResult.data,
```
Comment (sbddesign) [inline utils/posts.js:26]: See here @jordan-bravo: I added a default title and date, but these get overwritten by `...matterResult.data`.
  
Another option might be to just check if date or title is empty and kill the build process, spitting out a nice error message. But this works fine too.

**Pull request merged**